src/pybind/mgr/dashboard/services/nvmeof_client.py

```diff
@@ -13,18 +14,43 @@
 except ImportError:
     grpc = None
 else:
-    class NVMeoFClient(object):
+    class ChannelPool:
         def __init__(self):
-            logger.info('Initiating nvmeof gateway connection...')
-
             self.gateway_addr = list(NvmeofGatewaysConfig.get_gateways_config()[
                                      'gateways'].values())[0]['service_url']
-            self.channel = grpc.insecure_channel(
-                '{}'.format(self.gateway_addr)
-            )
+            self.pool_lock = threading.Lock()
+            self.channels = []
+
+        def get_channel(self):
+            with self.pool_lock:
+                if self.channels:
+                    return self.channels.pop()
+
+            return self.create_channel()
```

Comment:
I really like the idea and the implementation (and the fact that you're taking care of closing connections). I find, however, kinda conflicting that later thing with pooling, as usually with pools you want to avoid the overhead of creating/disposing resources.

That's usually for long-lasting usage (threads, connections, etc).

I'd go either path: short-lived connections with a context manager or long-lasting ones with pooling.